fuzzing/private/java_utils.bzl

```diff
@@ -90,17 +90,29 @@ def _java_segment_index(path_segments):
 
     return root_index
 
-def _jazzer_fuzz_binary_script(ctx):
-    script = ctx.actions.declare_file(ctx.label.name)
-
-    script_template = """
+_JAZZER_SCRIPT_TEMPLATE = """
 exec "{driver}" \
     --agent_path="{agent}" \
     --cp="{deploy_jar}" \
     --jvm_args="-Djava.library.path={library_path}" \
     "$@"
 """
 
+_JAZZER_OSS_FUZZ_SCRIPT_TEMPLATE = """#!/bin/sh
```

Comment:
The main differences are the following:

1. The fuzz target should run with the version of Jazzer provided by OSS-Fuzz, not the one it built from source itself.
2. Whereas fuzz targets should probably run with JVM defaults locally, OSS-Fuzz imposes stricter limits on memory usage that need to be followed by the JVM.
3. Java fuzz targets need to deviate from the default `ASAN_OPTIONS` in multiple ways as they use custom SEGV handlers and do not support leak detection.
4. Runfiles support is limited (flat directory instead of full runfile tree), meaning that paths need to be translated. This will always be necessary in some form as relative runfile paths such as `../target/deploy.jar` can never be realized in OSS-Fuzz.
5. ClusterFuzz runs fuzz targets with a different working directory, which requires the `$this_dir` dance.

It would be great if we could find ways to get rid of some of this complexity, but I can't see how we would be able to address all points transparently.

Reply:
Thanks for clarifying all these. It sounds good to me. Do you think we could define an OSS-Fuzz test project that (a) OSS-Fuzz developers can use as a starting point and (b) can be run as a Github action here through CI Fuzzing? I think we should have an end-to-end test demonstrating that all the pieces work together.

We do have a test project for the C++ rules at https://github.com/google/oss-fuzz/tree/master/projects/bazel-rules-fuzzing-test (CI Fuzzing integration [here](https://github.com/bazelbuild/rules_fuzzing/blob/master/.github/workflows/oss_fuzz.yml)), but I'm not sure the same project can build both C++ and Java fuzz targets at once.

CC @inferno-chromium @jonathanmetzman for their inputs too.